appcat/objectbucket/xrd.yaml

```diff
@@ -0,0 +1,50 @@
+apiVersion: apiextensions.crossplane.io/v1
+kind: CompositeResourceDefinition
+metadata:
+  name: xobjectbuckets.appcat.vshn.io
+spec:
+  group: appcat.vshn.io
+  names:
+    kind: XObjectBucket
+    plural: xobjectbuckets
+  claimNames:
+    kind: ObjectBucket
+    plural: objectbuckets
```

Comment:
I would actually not abstract away the platform (in this case cloudscale).

This allows the user to create buckets in a different cloud provider or region for offsite-backup purposes. For example, on Exoscale a customer might want to have the data in cloudscale or in a different exoscale datacenter, so that major incidents such as the data center fire at OVH Cloud in France doesn't impact the backups as well. With a platform-agnostic abstraction we lose this flexibility.
That doesn't mean that we support all kinds of providers and regions right from the start, but we wouldn't limit ourselves by choosing an API design that prevents this.
When rolling out on a cluster, e.g. on cloudscale, we could also rollout the composition for exoscale buckets so the users can choose.

The current implementation requires that the user needs to provide platform-specific details such as the region (so it's technically already platform-specific). If we also want to hide region from the user, then the composition should hardcode this to a value that is managed/configured by the Syn component.

@tobru WDYT? This idea [goes against the aspect](https://kb.vshn.ch/app-catalog/explanations/why_exists.html#_for_the_user) of "The developer who needs a service for the application but doesn’t want to be concerned about platform topics" but it keeps a business opportunity open for those users that actually have these concerns.

Alternatively, another composition could be made that is tailored towards for buckets just in the same region and cloud provider, but that is more to maintain and support, maybe even confusing for the user.
```suggestion
  group: cloudscale.appcat.vshn.io
  names:
    kind: XObjectBucket
    plural: xobjectbuckets
  claimNames:
    kind: ObjectBucket
    plural: objectbuckets
```
that means, eventually for Buckets on Exoscale, we'd have this:
```yaml
  group: exoscale.appcat.vshn.io # different group
  names:
    kind: XObjectBucket
    plural: xobjectbuckets
  claimNames:
    kind: ObjectBucket
    plural: objectbuckets
```

Reply:
I'm in favor of having just one XRD and abstract the details in the Composition. We set default values in the Composition for the specific cloud and configure the default Composition to use (`spec.defaultCompositionRef` in the XRD) depending on the location so that by default our well-chosen defaults are applied, the user then only has to care when a deviation from the default is wanted.

We can then deploy Compositions for Exoscale and cloudscale.ch on both (current) zones, configure the default Composition on cloudscale.ch to be the one for cloudscale.ch and on Exoscale to be the one for Exoscale. When a user on cloudscale.ch wants a bucket on Exoscale we instruct them (via documentation) how to choose another Composition. And if there is a wish to order a bucket on another zone but on the same provider, we instruct the user (again - via documentation) to set the corresponding parameter.

Reply:
We've had this discussion today again. Then I wasn't sure if the composition can be changed in the claim, but it looks like it's possible from the docs: https://crossplane.io/docs/v1.8/reference/composition.html#composite-resources-and-claims

In that case I'm also in favor of having one XRD with multiple compositions in the background.

Reply:
Ack, fine by me. However, just to reiterate, this assumes (and requires) that the user-facing API is compatible for all supported compositions.
I haven't tested the mechanism of specifying a different composition reference so I don't know if that's going to work as expected :shrug:

Reply:
I will leave the abstracted XRD as is.